subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertiesTaskStateChanges.java

```diff
@@ -16,41 +16,167 @@
+        private static final Equalizer<Object> DEFAULT_EQUALIZER = new DefaultEqualizer();
+        private final List<TaskStateChange> changes;
+
+        InputPropertyEqualizer(List<TaskStateChange> changes) {
+            this.changes = changes;
+        }
+
+        @Override
+        public boolean equals(@Nullable InputProperty obj1, @Nullable InputProperty obj2) {
+            if (obj1 == obj2) {
+                return true;
+            }
+            if (obj1 == null || obj2 == null) {
+                return false;
+            }
+
+            ClassLoader classLoader = selectClassLoader(obj1.getRawValue(), obj2.getRawValue());
+            Object raw1;
+            Object raw2;
+            try {
+                raw1 = tryDeserialize(obj1, classLoader);
+                raw2 = tryDeserialize(obj2, classLoader);
+            } catch (IOException e) {
+                // In presence of corruption, both object cannot assert to be equal
+                changes.add(new DescriptiveChange("Exception while deserializing the input property: %s", e.getMessage()));
+                return false;
+            } catch (ClassNotFoundException e) {
+                // In presence of complete change of property type, both object cannot assert to be equal
+                changes.add(new DescriptiveChange("Class definition couldn't be found while deserializing the input property: %s", e.getMessage()));
+                return false;
+            }
+
+            boolean isInputPropertyEquals = DEFAULT_EQUALIZER.equals(raw1, raw2);
+            if (!isEqualsMethodOverriden(raw1) || !isEqualsMethodOverriden(raw2)) {
+                changes.add(new DescriptiveChange("Input property is using the default equals method implementation which result in always detecting a change"));
+            } else {
+                maybeNagAboutCustomEqualsInTaskInputPropertyDeprecation(isInputPropertyEquals, Objects.equal(obj1.getHash(), obj2.getHash()));
+            }
+
+            return isInputPropertyEquals;
+        }
+
+        private static ClassLoader selectClassLoader(Object lhs, Object rhs) {
+            if (lhs == null && rhs == null) {
+                return InputProperty.class.getClassLoader();
+            } else if (lhs == null) {
+                return rhs.getClass().getClassLoader();
+            } else {
+                return lhs.getClass().getClassLoader();
+            }
+        }
+
+        private static Object tryDeserialize(InputProperty inputProperty, ClassLoader classLoader) throws IOException, ClassNotFoundException {
+            if (inputProperty.getRawValue() == null && inputProperty.getSerializedBytes() != null) {
+                return new ClassLoaderObjectInputStream(new ByteArrayInputStream(inputProperty.getSerializedBytes()), classLoader).readObject();
+            }
+
+            return inputProperty.getRawValue();
+        }
+
+        private static boolean isEqualsMethodOverriden(Object obj) {
```

Comment:
I think this is going to be a performance issue, as lookup in reflection is quite slow, and we're doing this check for both values for every input property of every task. We could either cache the result of the lookup, or remove the check (and just warn people that 'equals()' returns a different result to comparing the hashes).